### PR TITLE
[core][oom] enable group by parent policy by default

### PR DESCRIPTION
**Pull request by clarng**
## Why are these changes needed?

Fail the task if it is the last task of the group, per the new (group by parent) worker killing policy

## Related issue number

[32078](https://github.com/ray-project/ray/issues/32078)

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

Release test run : https://buildkite.com/ray-project/release-tests-pr/builds/26717
key tests to check for regression are under
core-daily-test
core-multi-test
core-nightly-test
data-tests
AIR tests

### CODE REVIEW

File: python/ray/tests/test_memory_pressure.py [hunk 1]
```diff
@@ -491,5 +491,38 @@ def test_legacy_memory_monitor_disabled_by_oom_killer():
...
+def test_last_task_of_the_group_fail_immediately():
+    @ray.remote(max_retries=-1)
+    def infinite_retry_task():
+        chunks = []
+        bytes_per_chunk = 1024 * 1024 * 1024
+        while True:
+            chunks.append([0] * bytes_per_chunk)
+            time.sleep(5)
+
+    # TODO(clarng): remove once these becomes the default
```
Comment (stephanie-wang) [inline python/ray/tests/test_memory_pressure.py:507]: Should we just do this in the same PR? It'd be good to get more test coverage.

**Comment by scv119** [comment]
@clarng do we have a list of (release)tests that we should look out for; and does all these release tests pass? it'd be nice to add that into the summary in this test.

File: src/ray/common/ray_config_def.h [hunk 2]
```diff
@@ -97,10 +97,10 @@ RAY_CONFIG(uint64_t, task_failure_entry_ttl_ms, 15 * 60 * 1000)
 /// ignored. This retry counter is only used when the process is killed due to memory, and
 /// the retry counter of the task or actor is only used when it fails in other ways
 /// that is not related to running out of memory. Retries indefinitely if the value is -1.
-RAY_CONFIG(uint64_t, task_oom_retries, 15)
+RAY_CONFIG(uint64_t, task_oom_retries, -1)
 
 /// The worker killing policy to use, as defined in worker_killing_policy.h.
```
Comment (scv119) [inline src/ray/common/ray_config_def.h:102]: should we document the list of policies here?

**Review by scv119 — APPROVED** [review]

**Comment by clarng** [comment]
is this good to merge? Will need to fix conflicts in https://github.com/ray-project/ray/pull/32107 asap once this is merged

@scv119  @stephanie-wang

**Pull request merged**